### DISCUSSION TITLE
How to use ts-check/jsdoc with javascript ava?

### DISCUSSION THREAD

**Discussion by sombriks** [avajs/ava#3338] [Q&A]
i know that there is type definitions but no luck on how to proper import it.

any guidance is welcome.

**Comment by manimovassagh** [comment]
You can get type checking with JSDoc in AVA test files. The key is importing AVA's types via JSDoc annotations.

### 1. Set up `jsconfig.json` (or `tsconfig.json` with `allowJs`)

```json
{
  "compilerOptions": {
    "checkJs": true,
    "strict": true,
    "module": "ESNext",
    "moduleResolution": "node",
    "target": "ESNext"
  },
  "include": ["**/*.js"]
}
```

### 2. In your test files, use JSDoc type imports

```js
import test from 'ava';

test('my test', (/** @type {import('ava').ExecutionContext} */ t) => {
  // t.is(), t.deepEqual(), etc. all have autocomplete now
  t.is(1 + 1, 2);
});
```

### 3. For typed context (like fixtures, beforeEach hooks)

```js
import anyTest from 'ava';

/** @type {import('ava').TestFn<{server: import('http').Server}>} */
const test = /** @type {any} */ (anyTest);

test.before((t) => {
  // t.context.server is typed here
});

test('uses server', (t) => {
  t.truthy(t.context.server);
});
```

The cast through `any` is needed when narrowing the generic type since JSDoc can't express generic function calls the way TypeScript can. But after the cast, you get full autocomplete and type checking on `t` and `t.context`.
